File: mouve/NodeTemporaryLinkView.cpp
#include "NodeTemporaryLinkView.h"
#include "NodeStyle.h"

NodeTemporaryLinkView::NodeTemporaryLinkView(const QPointF& startPosition,
	const QPointF& endPosition, QGraphicsItem* parent)
	: QGraphicsItem(parent)
	, mDrawDebug(false)
	, mDrawMode(1)
	, mPen(NodeStyle::LinkPen)
	, mEndPosition(mapFromScene(endPosition))
{
	setPos(startPosition);
	setZValue(NodeStyle::ZValueTemporaryLink);
}

void NodeTemporaryLinkView::paint(QPainter *painter,
	const QStyleOptionGraphicsItem* option, QWidget* widget)
{
	painter->setBrush(Qt::NoBrush);
	painter->setPen(mPen);
	painter->drawPath(_shape());
	// Draw debugging red-ish rectangle
	if(mDrawDebug)
	{
		painter->setBrush(QColor(255,0,0,50));
		painter->setPen(Qt::red);
		painter->drawRect(boundingRect());
	}
}

QRectF NodeTemporaryLinkView::boundingRect() const
{
	// !TODO: zbadac to
	float d = 0.0f;
	float x1 = qMin(qreal(0), mEndPosition.x()) - d;
	float y1 = qMin(qreal(0), mEndPosition.y()) - d;
	float x2 = qMax(qreal(0), mEndPosition.x()) + d;
	float y2 = qMax(qreal(0), mEndPosition.y()) + d;
	return QRectF(QPointF(x1, y1), QPointF(x2, y2));
}

void NodeTemporaryLinkView::updateEndPosition(const QPointF& endPosition)
{
	if(mEndPosition != endPosition)
	{
		prepareGeometryChange();
		mEndPosition = mapFromScene(endPosition);
	}
}

void NodeTemporaryLinkView::setDrawMode(int drawMode)
{
	mDrawMode = drawMode;
	update();
}

void NodeTemporaryLinkView::setDrawDebug(bool drawDebug)
{
	mDrawDebug = drawDebug;
	update();
}

QPainterPath NodeTemporaryLinkView::_shape() const
{
	// Default bezier drawing NodeStyle
	if(mDrawMode == 1)
	{
		QPointF c1(mEndPosition.x() / 2.0, 0);
		QPointF c2(mEndPosition.x() / 2.0, mEndPosition.y());

		QPainterPath painterPath;
		painterPath.cubicTo(c1, c2, mEndPosition);
		return painterPath;
	}
	//
	// Not quite working ;p 
	//
	else if(mDrawMode == 2)
	{
		qreal tangentLength = qAbs(mEndPosition.x()) / 2.0 +
			qAbs(mEndPosition.y()) / 4.0;
		QPointF startTangent(tangentLength, 0.0);
		QPointF endTangent(mEndPosition.x() - tangentLength, mEndPosition.y());

		QPainterPath painterPath;
		painterPath.cubicTo(startTangent, endTangent, mEndPosition);
		return painterPath;
	}
	// Simple line
	else
	{
		QPainterPath painterPath(QPointF(0, 0));
		painterPath.lineTo(mEndPosition);
		return painterPath;
	}	
}

File: UI/ViewPrototype/NodeTemporaryLinkView.py
from PySide.QtCore import *
from PySide.QtGui import *
from NodeStyle import *

'''
Used when user is dragging from one socket to another
'''
class NodeTemporaryLinkView(QGraphicsItem):
	def __init__(self, startPosition, endPosition, parent=None):
		QGraphicsItem.__init__(self, parent)

		self._drawDebug = False
		self._drawMode = 1
		self._pen = QPen(NodeStyle.LinkPen)

		# Origin is always at 0,0 in item coordinates
		self.setPos(startPosition)
		self.setZValue(NodeStyle.ZValueTemporaryLink)
		self._endPosition = self.mapFromScene(endPosition)

	def paint(self, painter, option, widget):
		painter.setPen(self._pen)
		painter.drawPath(self._shape())
		# Draw debugging red-ish rectangle
		if self._drawDebug:
			painter.setBrush(Qt.transparent)
			painter.setPen(Qt.red)
			painter.drawRect(self.boundingRect())

	def boundingRect(self):
		# !TODO: zbadac to
		d = 0.0
		x1 = min(0.0, self._endPosition.x()) - d
		y1 = min(0.0, self._endPosition.y()) - d
		x2 = max(0.0, self._endPosition.x()) + d
		y2 = max(0.0, self._endPosition.y()) + d
		return QRectF(QPointF(x1, y1), QPointF(x2, y2))

	@Slot(int)
	def setDrawMode(self, drawMode):
		self._drawMode = drawMode
		self.update()

	@Slot(bool)
	def setDrawDebug(self, drawDebug):
		self._drawDebug = drawDebug
		self.update()

	'''
	Buduje i zwraca QPainterPath 
	'''
	def _shape(self):
		endPos = self._endPosition

		# Default bezier drawing NodeStyle
		if self._drawMode == 1:
			c1 = QPointF(endPos.x() / 2.0, 0)
			c2 = QPointF(endPos.x() / 2.0, endPos.y())

			painterPath = QPainterPath()
			painterPath.cubicTo(c1, c2, endPos)
			return painterPath

		#
		# Not quite working ;p 
		#
		elif self._drawMode == 2:
			tangentLength = (abs(endPos.x()) / 2.0) + (abs(endPos.y()) / 4.0)
			startTangent = QPointF(tangentLength, 0.0)
			endTangent = QPointF(endPos.x() - tangentLength, endPos.y())

			painterPath = QPainterPath()
			painterPath.cubicTo(startTangent, endTangent, endPos)
			return painterPath

		# Simple line
		else:
			painterPath = QPainterPath(QPointF(0, 0))
			painterPath.lineTo(endPos)
			return painterPath

	'''
	Uaktualnia pozycje koncowa
	'''
	def updateEndPosition(self, endPosition):
		if self._endPosition != endPosition:
			self.prepareGeometryChange()
			self._endPosition = self.mapFromScene(endPosition)

	#
	# type property
	#

	Type = QGraphicsItem.UserType + 5

	def type(self):
		return NodeTemporaryLinkView.Type